### PR TITLE
Switch clang-tidy plugin and workflow to LLVM 12

### PR DESCRIPTION
**Pull request by jbytheway**
#### Summary
Infrastructure "Switch clang-tidy plugin and workflow to LLVM 12"

#### Purpose of change
Our clang-tidy workflow is using LLVM 8.  That's three major versions behind now, so it seems a good time to update to the latest version, and take advantage of all the new checks that have been added.

#### Describe the solution
* Elsewhere on my personal repo, released a new version of [clang-tidy-plugin-support](https://github.com/jbytheway/clang-tidy-plugin-support/releases/tag/llvm-12.0.0-r1).
* Use that new version in `CMakeLists.txt`.
* Port the plugin code to the new LLVM version (mostly this just meant adding a bunch of header includes, but a couple of function signatures had changed too).
* Update the GitHub workflow file.
* Disable all the new checks which warn on our code.

#### Describe alternatives you've considered
Could have fixed some of the checks' warnings instead of simply disabling them all, but this seemed

#### Testing
Running clang-tidy locally and using CI here.  Testing is ongoing while this is a draft PR.

#### Additional context
Post 0.F, I plan to update all our CI as a result of Ubuntu Xenial going out of support.  This seemed like a low-risk place to start that work.

This PR was originally for LLVM 11, but switched to 12 once it was pointed out that that had released.

### CODE REVIEW

**Comment by BrettDong** [comment]
Is there a reason not using LLVM 12?

**Comment by jbytheway** [comment]
> Is there a reason not using LLVM 12?

There hadn't been a stable release of 12 when I started thinking about this.  It was less than two weeks ago.  My experience with first releases of new gcc major versions has not been that great, so I'm a little scared to use 12.0.0, but maybe clang is more reliable.  It probably makes sense to use 12.  We move to 12.0.1 if necessary when it comes out with fixes.

**Comment by Qrox** [comment]
`dev/DEVELOPER_TOOLING.md` still refers to llvm 8 and needs to be updated too.

**Comment by jbytheway** [comment]
> `dev/DEVELOPER_TOOLING.md` still refers to llvm 8 and needs to be updated too.

Thanks for the reminder.,  I've updated it now (although the Windows part is presumably now wrong).

**Comment by Qrox** [comment]
> > `dev/DEVELOPER_TOOLING.md` still refers to llvm 8 and needs to be updated too.
> 
> Thanks for the reminder., I've updated it now (although the Windows part is presumably now wrong).

I've pushed the updated docs for Windows and created a PR to your repository (https://github.com/jbytheway/Cataclysm-DDA/pull/3).

**Comment by jbytheway** [comment]
OK, I think this is now good.  Marked ready for review.

**Review by Qrox — COMMENTED** [review]
Update docs after changes to https://github.com/jbytheway/clang-tidy-plugin-support/pull/3.

**Comment by kevingranade** [comment]
Reminder that #48802 either needs to be reverted or not merged to -dev along with this.

File: doc/DEVELOPER_TOOLING.md [hunk 1]
```diff
@@ -66,7 +66,7 @@ In addition to the usual means of creating a `tags` file via e.g. [`ctags`](http
 
 Cataclysm has a [clang-tidy configuration file](../.clang-tidy) and if you have
 `clang-tidy` available you can run it to perform static analysis of the
-codebase.  We test with `clang-tidy` from LLVM 8.0.1 on Travis, so for the most
+codebase.  We test with `clang-tidy` from LLVM 12.0.0 on CI, so for the most
```
Comment (actual-nh) [inline doc/DEVELOPER_TOOLING.md:69]:
```suggestion
codebase.  We test with `clang-tidy` from LLVM 12.0.0 with CI, so for the most
```

File: doc/DEVELOPER_TOOLING.md [hunk 2]
```diff
@@ -66,7 +66,7 @@ In addition to the usual means of creating a `tags` file via e.g. [`ctags`](http
...
 consistent results, you might want to use that version.
 
 To run it you have a few options.
```
Comment (actual-nh) [inline doc/DEVELOPER_TOOLING.md:72]:
```suggestion
To run it, you have a few options.
```

File: doc/DEVELOPER_TOOLING.md [hunk 3]
```diff
@@ -92,12 +92,12 @@ We have written our own clang-tidy checks in a custom plugin.  Unfortunately,
...
+#### Ubuntu Focal
 
-If you are on Ubuntu Xenial then you might be able to get it working the same
-way Travis does.  Add the LLVM 8 Xenial source [listed
-here](https://apt.llvm.org/) to your `sources.list`, install the `clang-8
-libclang-8-dev llvm-8-dev llvm-8-tools` packages, and build Cataclysm with CMake,
+If you are on Ubuntu Focal then you might be able to get it working the same
+way CI does.  Add the LLVM 12 Focal source [listed
+here](https://apt.llvm.org/) to your `sources.list`, install the `clang-12
+libclang-12-dev llvm-12-dev llvm-12-tools` packages, and build Cataclysm with CMake,
```
Comment (actual-nh) [inline doc/DEVELOPER_TOOLING.md:100]:
```suggestion
way our CI does.  Add the LLVM 12 Focal source [listed
here](https://apt.llvm.org/) to your `sources.list`, install the needed packages (`clang-12
libclang-12-dev llvm-12-dev llvm-12-tools`), and build Cataclysm with CMake,
```

File: doc/DEVELOPER_TOOLING.md [hunk 4]
```diff
@@ -151,14 +150,13 @@ After the tools are installed, a patch still needs to be applied before building
 llvm, since `clang-tidy` as distributed by LLVM doesn't support plugins.
 
 First, clone the llvm repo from for example [the official github repo](https://github.com/llvm/llvm-project.git).
```
Comment (actual-nh) [inline doc/DEVELOPER_TOOLING.md:152]:
```suggestion
First, clone the llvm repo from, for example, [the official github repo](https://github.com/llvm/llvm-project.git).
```

File: doc/DEVELOPER_TOOLING.md [hunk 5]
```diff
@@ -151,14 +150,13 @@ After the tools are installed, a patch still needs to be applied before building
...
-Checkout the `release/8.x` branch, since that's where our patch was based on.
+Checkout the `release/12.x` branch, since that's where our patch was based on.
```
Comment (actual-nh) [inline doc/DEVELOPER_TOOLING.md:153]:
```suggestion
Checkout the `release/12.x` branch, since that's what our patch was based on.
```

**Review by actual-nh — COMMENTED** [review]
Just a few grammar/style nits.

**Comment by jbytheway** [comment]
Another reminder that we may want to revert #49000 after this is merged.

**Pull request merged**